### PR TITLE
MAJOR: Change internal encoding for Task.Async

### LINKED ISSUE #612: TaskLocal doesn't preserve locals for Task.executeOn
See code samples below.

`asyncBoundary` works fine:
```scala
implicit val ec: Scheduler = monix.execution.Scheduler.Implicits.global
val ec2: Scheduler = monix.execution.Scheduler.io()
implicit val opts = Task.defaultOptions.enableLocalContextPropagation

for {
  local <- TaskLocal(0).asyncBoundary(ec2)
  _ <- local.write(100).asyncBoundary(ec)
  v1 <- local.read.asyncBoundary(ec2) // 100
  v2 <- local.read // 100
  _ <- Task { println(s"$v1 : $v2")}
} yield ()
```

`executeOn`:
```scala
for {
  local <- TaskLocal(0)
  _ <- local.write(100).executeOn(ec2)
  v1 <- local.read.executeOn(ec)  // 0
  v2 <- local.read.executeOn(ec2) // 0
  _ <- Task { println(s"$v1 : $v2")}
} yield ()
```

[Link to the discussion on Gitter](https://gitter.im/monix/monix?at=5aafaafb6f8b4b99464384fb)

### PR DESCRIPTION
**Pull request by alexandru**
Fixes #612 — The initial implementation for `TaskLocal`, which I contributed after collaborated with @leandrob13 on `Local` and on `TracingSchedule`, was a little superficial as it did not take into account `executeOn` among others.

So upon defining `Task.Async` like ...

```scala
Task.Async { (ctx, cb) =>
  ctx.scheduler.executeAsync(() => cb.onSuccess(1))
}
```

What the (current) implementation does is to read the current locals context via `Local.getContext` and after this task gets evaluated restore the context via `Local.setContext`, which happens upon calling `cb.onSuccess`. The virtue of this implementation is that it doesn't matter if the `ctx.scheduler` is a `TracingScheduler` or not.

Well, that's not good enough for operations like `executeOn`, operations that in fact wrap another `Task`:

```scala
Async { (ctx: Context, cb: Callback[A]) => 
  val ctx2 = ctx.withScheduler(s)
  if (forceAsync)
   Task.unsafeStartAsync(source, ctx2, cb)
  else
    Task.unsafeStartNow(source, ctx2, cb)
}    
```

Restoring the previous context in such a case isn't good, because the task that we are executing here could apply changes to the local context that we want to be available in bind chains after the evaluation happens.

To fix this I introduced 2 changes:

1. the `Scheduler` passed in `Async` tasks via `Context` will now be a `TracingScheduler` in case `localContextPropagation` is on
2. I changed the encoding of `Async` to allow disabling this restoration of the context for operations that wrap other tasks, like `executeOn` is doing

The new definition is:

```scala
final case class Async[+A](
    register: (Context, Callback[A]) => Unit,
    trampolineBefore: Boolean = false,
    trampolineAfter: Boolean = true,
    restoreLocals: Boolean = true)
    extends Task[A]
```

If `restoreLocals` is `false`, then we are not restoring the `Local.Context` after evaluation — this will now make `executeOn` behave like expected.

---

While working on this I also noticed that `RestartCallback` admits an optimization for triggering asynchronous boundaries. So asynchronous boundaries in `Task.Async` are mandatory, except that:

- sometimes we need them only on the callback, so after evaluation
- sometimes we need them before evaluating Async's function
- sometimes we need async boundaries both before and after evaluation

Therefore I introduced two booleans in `Async` such that:

1. if `trampolineBefore` is `true`, then it triggers an optimized async boundary before evaluation
2. if `trampolineAfter` is `true`, it triggers an async boundary after evaluation, but before calling the callback (on by default)

Before this we had to do ...

```scala
def signal(n: Int): Task[Int] = 
  Task.Async { (ctx, cb) =>
    // Obligatory
    cb.asyncOnSuccess(n)(ctx.scheduler)
  }
```

After this change we can do:

```scala
def signal(n: Int): Task[Int] = 
  Task.Async(
    (_, cb) => cb.onSuccess(n),
    trampolineAfter = true)
```

With this change the memory usage becomes more efficient because in managing these boundaries the `RestartCallback` implementation can avoid creating extraneous `Runnable` instances.

---

As part of this task it became clear that `Task.unsafeCreate` is really unsafe. This is why we're doing a deprecations:

1. `Task.unsafeCreate` is now deprecated, to be removed at a later time
2. `Task.async` is also deprecated, because it doesn't reflect `cats.effect.Async`

In place of `unsafeCreate` I've expanded the available builders:

1. `Task.async` is now the equivalent of `cats.effect.Async#async`; which is cool to have in place of using `create` with a `Cancelable.empty` return because it's more optimal (we don't have to do the internal cancelable bookkeeping)
2. `Task.asyncS` is the `async` builder that also injects a `Scheduler` (`S` suffix comes from `Scheduler`; naming is hard 😑)
3. `Task.cancelable` is the equivalent of `cats.effect.Cancelable#cancelable`
4. `Task.cancelableS` is the `cancelable` builder that also injects a `Scheduler` (just like `async` vs `asyncS`)
3. `Task.create` has been changed to use the [Partially-Applied Type technique](https://typelevel.org/cats/guidelines.html#partially-applied-type-params) and thus to change behavior depending on the return type of the provided function ... it supports `Unit` (equivalent with `Task.simple` / `async`), it also supports `Cancelable`, `IO[Unit]`, `Coeval[Unit]` and `Task[Unit]` as return types

---

After changing the `Async` implementation I had to review and fix the implementation of the following:

- [x] executeOn
- [x] executeWithModel
- [x] executeWithOptions
- [x] deferAction
- [x] uncancelable
- [x] raiseError
- [x] fromFuture / deferFutureAction
- [x] fromAsync / fromConcurrent
- [x] simple / cancelable
- [x] doOnCancel
- [x] startAndForget
- [x] gather
- [x] gatherUnordered
- [x] mapBoth
- [x] memoize
- [x] memoizeOnSuccess
- [x] race
- [x] raceList
- [x] racePair
- [x] sleep
- [x] start

N.B. `TaskLocal` due to its behavior might still yield surprises. For example `Task.gather` executes tasks in parallel, so if any of those tasks does a context `write`, then that `write` cannot get propagated after `gather` is finished. Same goes for `start`, for the same reasons.

Basically when concurrent execution is involved (e.g. `mapBoth`, `race`, etc.) then we cannot keep the changes to the local context in the bind continuation.

### CODE REVIEW

**Comment by Avasil** [comment]
> Should we make unsafeCreate private? I kind of think so — we can also add extra create overloads for optimization purposes and that should be fine for most use-cases; and if anybody needs unsafeCreate, he/she should try to push the use-case in Monix proper

Let's make it private - I'm not a fan of public builder that we advise against using at every step. Still, we need to have in mind people who might had to use it for certain use cases and this change might prevent their upgrade from `2.3.x` to `3.x.x`. If we can't cover all of the use cases now maybe we can deprecate it in `3.0.0` and remove in `3.1.0`? So it's not that sudden.

> TracingScheduler only takes care of real async boundaries, not trampolined ones — should it also save and restore the context for trampolined execution as well (e.g. TrampolinedRunnable)? At this point I do not think so, but I might be wrong

What are repercussions of not restoring it?

> I'm thinking of making an issue / document with all use-cases for TaskLocal we can think of, introducing necessary tests and making sure that this isn't missing anything.

Yes please, it will not only help with assuring that `TaskLocal` implementation is correct but we could also easily integrate it into documentation / blog post. I think it's not instantly obvious why `TaskLocal` is awesome so stuff like this will be very useful.

**Comment by leandrob13** [comment]
1.- I agree unsafeCreate should be private

2.- We can keep beta testing to see if there is a use case where the propagation is needed in trampolined execution.

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala [hunk 1]
```diff
@@ -3115,12 +3370,20 @@ object Task extends TaskInstancesLevel1 {
...
+    * @param register is the side-effecting, callback-enabled function
+    *        that starts the asynchronous computation and registers
+    *        the callback to be called on completion
     */
-  private[eval] final case class Async[+A](register: (Context, Callback[A]) => Unit)
+  private[eval] final case class Async[+A](
+    register: (Context, Callback[A]) => Unit,
+    trampolineBefore: Boolean = false,
+    trampolineAfter: Boolean = true,
+    restoreLocals: Boolean = true)
```
[the remark this answers is not in the document]
Comment (alexandru) [reply monix-eval/shared/src/main/scala/monix/eval/Task.scala]: Well, it's an internal thing, it can be changed later if we'll need it.

File: monix-eval/shared/src/main/scala/monix/eval/Task.scala [hunk 2]
```diff
@@ -2107,57 +2105,196 @@ object Task extends TaskInstancesLevel1 {
...
+    * injecting a [[monix.execution.Scheduler]] for us, which we can
+    * use as an `ExecutionContext` for executing `Future#onComplete`.
     *
-    * @param register $registerParamDesc
+    * @see [[Task.cancelable[A](start* Task.cancelable]] for 
+    *      creating cancelable tasks
     */
-  def async[A](register: (Scheduler, Callback[A]) => Cancelable): Task[A] =
-    create(register)
+  def simple[A](start: (Scheduler, Callback[A]) => Unit): Task[A] =
```
[the remark this answers is not in the document]
Comment (alexandru) [reply monix-eval/shared/src/main/scala/monix/eval/Task.scala]: Another possibility would be to have `Task.async` with a `Function1` parameter, as you are suggesting, then `async0` or `async2` (or some other name that's not `async`) for the `Function2`. But then we'll need to do it for `cancelable` as well, for consistency.

So what kind of overload convention would you prefer?

[the remark this answers is not in the document]
Comment (oleg-py) [reply monix-eval/shared/src/main/scala/monix/eval/Task.scala]: @alexandru is new `Task.create` not a such builder? I thought that is a strictly more generic version, just return a `Unit` if you want it uncancelable, and existing code can migrate to it directly.

[the remark this answers is not in the document]
Comment (oleg-py) [reply monix-eval/shared/src/main/scala/monix/eval/Task.scala]: Crazy idea: `def async(f: CallbackWithScheduler[A] => Unit)`

where f.scheduler would be implicitly available, so you don't even have to do
```scala
implicit val sc = scheduler
```

Less crazy idea, but probably less efficient: `Scheduler => Callback[A] => Unit`, lets you do `implicit sc => cb => { ... }`

**Comment by alexandru** [comment]
@oleg-py I renamed `simple` to `async` and introduced `asyncS` and `cancelableS` (`S` suffix comes from `Scheduler`).

I don't have better names and they need to be exposed. In absence of better options, we're going to go with these.

**Comment by oleg-py** [comment]
Works for me 👍
